hazelcast-sql-core/src/main/java/com/hazelcast/sql/impl/calcite/validate/types/HazelcastTypeCoercion.java

```diff
@@ -79,15 +79,13 @@ public boolean coerceOperandType(SqlValidatorScope scope, SqlCall call, int inde
     private boolean requiresCast(SqlValidatorScope scope, SqlNode node, RelDataType to) {
         RelDataType from = validator.deriveType(scope, node);
 
-        assert from.isNullable() == to.isNullable();
-
         if (from.getSqlTypeName() == NULL || SqlUtil.isNullLiteral(node, false) || node.getKind() == SqlKind.DYNAMIC_PARAM) {
             // Never cast NULLs or dynamic params, just assign types to them
             return false;
         }
 
         // CAST is only required between different types.
-        return from.getSqlTypeName() != to.getSqlTypeName();
+        return !from.equals(to);
```

Comment:
This change is incorrect. We add `CAST` only when type names are different. If type names are equal, but types are different (e.g. `BIGINT(63)` vs `BIGINT(64)`), we only update the operand type in the validator. See usages of `HazelcastTypeCoercion.coerceOperandType`.